### ISSUE TITLE
pathmap command line option inconsistent with similar command line options such as link/reference with regards to paths containing spaces

### ISSUE THREAD

**Issue by jeffanders** [dotnet/roslyn#19829]
Pathmap is fairly new and as of yet undocumented compiler command line option (see #18071 and dotnet/docs#1800). The main purpose of pathmap is to remap paths of source files for the benefit of deterministic builds that use CallerFilePathAttribute.

The general format of the pathmap command line option is:
/pathmap:path1=pathA,path2=pathB etc

If any of the paths provided to the pathmap parameter contain spaces then you need to put double quotes around the entire set of comma separated path mapping instead of being able to quote individual paths. 

ParsePathMap is inconsistent with other command line options that take a comma separated list of files or paths. For example the link and reference command line options make use of ParseAssemblyReferences which gives the following examples in comments within that method:

// /r:"reference"
// /r:alias=reference
// /r:alias="reference"
The MSDN documentation for the link option (https://msdn.microsoft.com/en-us/library/dd264728.aspx) explicitly states "If the file name contains a space, enclose the name in quotation marks".

It is expected that pathMap should work in a similar way.

**Version Used**: 
This is using latest Master. I actually came across this while updating all Roslyn tests to be compatible with user accounts/temp paths containing spaces.  Please refer to pull request #17524 for further details.

**Steps to Reproduce**:

1. Create a new folder containing a space, for example "c:\path with spaces" and create a program.cs file in that folder with the following contents:
``` C#
using System;

class Program
{
    static void Main(string[] args) => TraceMessage();
    static void TraceMessage([System.Runtime.CompilerServices.CallerFilePath] string sourceFilePath = "") => Console.WriteLine("source file path: " + sourceFilePath);
}
```
2. Compile the project from the command line as follows:
```csc.exe program.csc /pathmap:"c:\path with spaces"=\ ```
3. Run program.exe

**Expected Behavior**:
Produce a program.exe that when run produces the following output:
```source file path: c:\path with spaces\program.cs"```

**Actual Behavior**:
csc.exe returns an error because the command line was malformed with everything from the word ```with``` onwards being treated as extra separate command line options. The error returned is:
```error CS8101: The pathmap option was incorrectly formatted.```

To successfully compile this with the pathmap you must use the following command line (note the closing double quote is now after the forward slash after the equals sign):
```csc.exe program.csc /pathmap:"c:\path with spaces=\" ```

**Comment by jcouv** [comment]
FYI @jaredpar

**Comment by jaredpar** [comment]
> The main purpose of pathmap is to remap paths of source files for the benefit of deterministic builds that use CallerFilePathAttribute.

There are actually a few reasons for it:

1. Remap paths used in `CallerFilePathAttribute`
1. Change the path of the source files as embedded in the PDB 
1. Change the path of the PDB embedded into the PE 


At a glance though I can't remember exactly why we differentiated here.  @TyOverby or @gafter should know.

**Comment by jcouv** [comment]
Ping @TyOverby @gafter

**Comment by gafter** [comment]
I don't think the difference was intentional.

**Comment by jeffanders** [comment]
If this is something that should be changed to be consistent, but is a low priority, I am willing to work on this. I would probably just need to know when a pull request would be needed by.  Presumably this would need to be done before the docs are written for pathMap (the docs issue just got updated with a milestone of July 2017).

**Comment by GuardRex** [comment]
@jeffanders I'll be handling the doc. I was just taking a look at [the doc issue](https://github.com/dotnet/docs/issues/1800) this evening (6/26). It would be helpful if `/pathmap` is stable; however, we can cover it in its present form and open an issue blocked by this issue for updates later.

As you say, the deciding factor should be *when* would the updates to `/pathmap` occur. If they would occur within a few weeks, then I suggest we hold off a sec until it's stable. If it's more like in the months horizon, then I suggest we go ahead with the documentation now. Thoughts?

**Comment by jeffanders** [comment]
@GuardRex, I've got a bit happening over the next couple of weeks but would like to get this done.

This originated with me working on a roslyn fork where I implemented about a third of a proposal for ! based non nullable references. After a merge from Roslyn tests were broken horribly so I ceased work on it for a year hoping someone would fix them. No one did so eventually I just worked through all the broken tests, realising they related to a space in my account name, and created a pull request. But the determinism tests were still failing. That led me to the undocumented pathMap option so i fixed the tests and logged this issue. I'm not going back to my original fork (it is clear the ldt is going with ? based nullable references)

So I think I would like to finish shaving this yak if not for my own mental well being of actually having finally achieved something :stuck_out_tongue_winking_eye:.

How about If I can submit a pull request or provide a solid update in the next fortnight defer the doc updates otherwise proceed as normally planned.

**Comment by GuardRex** [comment]
@jeffanders I can go either way with the doc coverage.

I can doc it as it currently sits probably the end of next week (~July 7) and come back later. The original request was made some time ago on March 24, and it was just assigned to me. I'm anxious to deliver on this.

However, it doesn't make much sense to me to cover it ~7/7 if the following week (by 7/14) the syntax updates are made and then more doc work has to be scheduled another week or two beyond that.

If you find that it's going to take longer ... if the updates are going to run out beyond ~7/14, then ping me here, and I'll go ahead the end of next week. I'll also file a new issue on the docs repo to do updates beyond 7/7 and block that issue on this one so that we get back to it later.

**Comment by gafter** [comment]
Dup of #22835. (Deferring to the more recent assignment).
